compiler/src/iree/compiler/Codegen/Common/Transforms.cpp

```diff
@@ -354,6 +355,108 @@ void populateSwapExtractWithExpandPattern(RewritePatternSet &patterns) {
   patterns.add<SwapExpandShapeWithSlicePattern>(patterns.getContext());
 }
 
+namespace {
+
+/// Pattern to fold extract_slice(broadcast) into the broadcast input when the
+/// extract_slice is rank-reducing and only extracts along the broadcasted
+/// dimensions (with size 1), leaving all non-broadcasted dimensions intact.
+/// This is valid because the broadcast only duplicates data along the
+/// broadcasted dimension, and extracting a single slice from that dimension
+/// gives back the original input.
+///
+/// Example:
+///   %broadcast = linalg.broadcast ins(%in : tensor<4x1xf16>)
+///                                 outs(%out : tensor<4x1x1xf16>) dimensions =
+///                                 [2]
+///   %extract = tensor.extract_slice %broadcast[0, 0, 0] [4, 1, 1] [1, 1, 1]
+///              : tensor<4x1x1xf16> to tensor<4x1xf16>
+/// ->
+///   %extract is replaced by %in (tensor<4x1xf16>)
+struct FoldExtractSliceOfBroadcast final
+    : OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    auto broadcastOp =
+        extractOp.getSource().getDefiningOp<linalg::BroadcastOp>();
+    if (!broadcastOp) {
+      return rewriter.notifyMatchFailure(
+          extractOp, "source is not a linalg.broadcast operation");
+    }
+
+    if (!extractOp.hasUnitStride()) {
+      return rewriter.notifyMatchFailure(extractOp,
+                                         "extract_slice has non-unit stride");
+    }
+
+    auto inputType =
+        dyn_cast<RankedTensorType>(broadcastOp.getInput().getType());
+    auto broadcastOutputType =
+        dyn_cast<RankedTensorType>(broadcastOp.getInit().getType());
+    auto extractResultType =
+        dyn_cast<RankedTensorType>(extractOp.getResult().getType());
+    if (!inputType || !broadcastOutputType || !extractResultType) {
+      return rewriter.notifyMatchFailure(
+          extractOp, "operand or result types are not RankedTensorType");
+    }
+
+    // Extract result type must match broadcast input type.
+    if (inputType != extractResultType) {
+      return rewriter.notifyMatchFailure(
+          extractOp, "extract result type does not match broadcast input type");
+    }
+
+    // Verify that we're extracting from offset [0, 0, ..., 0] with the same
+    // shape as the input (essentially undoing the broadcast).
+    SmallVector<OpFoldResult> offsets = extractOp.getMixedOffsets();
+    SmallVector<OpFoldResult> sizes = extractOp.getMixedSizes();
+    if (llvm::any_of(offsets, [](OpFoldResult offset) {
+          return !isConstantIntValue(offset, 0);
+        })) {
```

Comment:
```suggestion
    if (!llvm::all_of(offsets, isZeroInteger)) {
```
I think this aligns better with the match failure message.

(I realize this code is moved over, so fine to keep as-is in this PR)